otwcode/otwarchive
# AO3-6233 Highlight autocomplete suggestions with diacritics

# Pull Request Checklist

  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6233

## Purpose

Fixes the autocomplete highlighting to handle diacritic marks so that matching substrings in suggestions are bolded regardless of diacritic marks.  Makes the changes from #4009 more obvious to the user.

<img width="637" alt="Screen Shot 2021-09-08 at 12 16 56 AM" src="https://user-images.githubusercontent.com/6746746/132445833-69e4174c-cf74-4109-b254-0b128445708e.png">
<img width="645" alt="Screen Shot 2021-09-08 at 12 17 39 AM" src="https://user-images.githubusercontent.com/6746746/132445835-0e41ae9b-841d-4e1d-ba92-30fb81d7fc4e.png">

## Testing Instructions

Cucumber tests have been added to `autocomplete.feature`.

For manual testing:
- Create canonical tags of various types, both with and without diacritics
- Reload the tag data for the autocomplete `rake autocomplete:reload_tag_data`
- Go to the form to post a new work (or edit an existing one)
- Enter some values in the tag fields and ensure the autocomplete brings up suggestions with the appropriate parts highlighted.

The updated method for highlighting could be slow if there are a lot of long potentially matching tags.  Performance seems reasonable to me, but that might be dependent on device, and I did not run a test with more than a few tag suggestions at a time.

I've also incidentally changed a behaviour which I think was unintentional before.  Now only the first occurence of match in a suggestion will be highlighted.  For example, if you type `f` and "Fluff" is in the suggestions, only the first "F" will be bolded, instead of all three.  I think this makes more sense as the autocomplete is only keying in on matches at the start of words anyway.

## References

https://otwarchive.atlassian.net/browse/AO3-2380
#4009
